### ISSUE TITLE
Ansible 2.0.0.2 skipped tasks with with_nested does not handle undefined variables

### ISSUE THREAD

**Issue by davidfischer-ch** [ansible/ansible#14498]
Issue Type : Bug Report
Ansible Version: 2.0.0.2
Environment: Ubuntu 14.04 LTS Python 2.7.10
Summary:

I have a bunch of tasks that are skipped on some condition. One of them set_fact a variable. Even if the task using with_nested is skipped, the with_nested argument tries to "access" the variable and fail.

I tried to create a block and put all tasks with the same when condition into the block but the issue is the same.

On ansible 1.9.4 it works flawless.

Steps To Reproduce:

```
- set_fact: something=[1, 2]
  when: no

- name: Fail because of this issue
  command: echo {{ item[0] }} {{ item[1] }}
  when: no
  with_nested:
    - "{{ something }}"
    - [3, 4]
```

Excepted : all tasks are skipped
Actual:

`fatal: [my-pc]: FAILED! => {"failed": true, "msg": "ERROR! 'something' is undefined"}`

Regards,

David

**Comment by jimi-c** [comment]
Hi @davidfischer-ch, we've actually corrected this issue, however the result is not quite what you'd think:

<pre>
TASK [set_fact] ****************************************************************
task path: /root/testing/14498.yml:4
skipping: [localhost] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
TASK [Fail because of this issue] **********************************************
task path: /root/testing/14498.yml:7
[DEPRECATION WARNING]: Skipping task due to undefined Error, in the future this will be a fatal error.: 'something' is undefined. This feature will be removed in a future release. 
Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
fatal: [localhost]: FAILED! => {"failed": true, "msg": "with_nested requires at least one element in the nested list"}
</pre>

In this case, the error is specific to with_nested due to the way it's not expanding that one list. So it looks like we'll have to add a little extra consideration for situations like this.

**Comment by prupert** [comment]
We are still experiencing this issue with ansible 2.0.1.0 (RPM package, stable from EPEL7).

**Comment by prupert** [comment]
I am just wondering if there is any progress on fixing this issue?

**Comment by jimi-c** [comment]
Hi @prupert, it appears this issue has been resolved in both stable-2.1 and devel. On both, the output is:

<pre>
$ ansible-playbook -vv 14498.yml 
Using /etc/ansible/ansible.cfg as config file
PLAYBOOK: 14498.yml ************************************************************
1 plays in 14498.yml
PLAY [localhost] ***************************************************************
TASK [set_fact] ****************************************************************
task path: /home/jamesc/testing/14498.yml:4
skipping: [localhost] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
TASK [Fail because of this issue] **********************************************
task path: /home/jamesc/testing/14498.yml:7
[DEPRECATION WARNING]: Skipping task due to undefined Error, in the future this will be a fatal error.: 'something' is undefined.
This feature will be removed in a future release. 
Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
skipping: [localhost] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=0   
</pre>

It looks like this has been the case since at least 2.1.0, so we'll go ahead and close this now.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
